src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -587,24 +587,53 @@ std::wstring Terminal::GetHyperlinkAtViewportPosition(const til::point viewportP
 
 std::wstring Terminal::GetHyperlinkAtBufferPosition(const til::point bufferPos)
 {
-    auto attr = _activeBuffer().GetCellDataAt(bufferPos)->TextAttr();
+    // Case 1: buffer position has a hyperlink stored in the buffer
+    const auto attr = _activeBuffer().GetCellDataAt(bufferPos)->TextAttr();
     if (attr.IsHyperlink())
     {
-        auto uri = _activeBuffer().GetHyperlinkUriFromId(attr.GetHyperlinkId());
-        return uri;
+        return _activeBuffer().GetHyperlinkUriFromId(attr.GetHyperlinkId());
     }
-    // also look through our known pattern locations in our pattern interval tree
-    auto viewportPos = bufferPos;
-    _GetVisibleViewport().ConvertToOrigin(&viewportPos);
-    const auto result = GetHyperlinkIntervalFromViewportPosition(viewportPos);
+
+    // Case 2: buffer position may point to an auto-detected hyperlink
+    // Case 2 - Step 1: get the auto-detected hyperlink
+    std::optional<interval_tree::Interval<til::point, size_t>> result;
+    const auto visibleViewport = _GetVisibleViewport();
+    if (visibleViewport.IsInBounds(bufferPos))
+    {
+        // Hyperlink is in the current view, so let's just get it
+        auto viewportPos = bufferPos;
+        visibleViewport.ConvertToOrigin(&viewportPos);
+        result = GetHyperlinkIntervalFromViewportPosition(viewportPos);
+        if (result.has_value())
+        {
+            result->start = _ConvertToBufferCell(result->start);
+            result->stop = _ConvertToBufferCell(result->stop);
+        }
+    }
+    else
+    {
+        // Hyperlink is outside of the current view.
+        // We need to find if there's a pattern at that location.
+        const auto patterns = _activeBuffer().GetPatterns(bufferPos.y, bufferPos.y);
+
+        // NOTE: patterns is stored with top y-position being 0,
+        //       so we need to cleverly set the y-pos to 0.
+        const til::point viewportPos{ bufferPos.x, 0 };
+        const auto results = patterns.findOverlapping(viewportPos, viewportPos);
+        if (!results.empty())
+        {
+            result = results.front();
+            result->start.y += bufferPos.y;
+            result->stop.y += bufferPos.y;
+        }
+    }

```

Comment:
Why do we have to handle hyperlinks outside of the viewport?

Reply:
Keyboard-based navigation through the buffer looking for links would not be very effective if it was constrained to only working in the visible region :)

Reply:
Oh I guess the scrolling is happening in the caller then?

Reply:
I guess I didn't think about how this function was used. It's GOOD that it is robust now, but I suppose now that it was never _called_ with coordinates outside the viewport, eh?

Reply:
I think given due to https://github.com/microsoft/terminal/issues/13854#issuecomment-1233590998, long-term, we should probably have everything be in the buffer coordinate system. It'll simplify a lot of this extra code. We also need to find hyperlinks outside of the existing viewport. I assume we didn't do that the first time due to performance concerns. :/